plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/ThriftHiveMetastoreClient.java

```diff
@@ -244,13 +257,25 @@ public void deleteTableColumnStatistics(String databaseName, String tableName, S
     }
 
     @Override
-    public void setPartitionColumnStatistics(String databaseName, String tableName, String partitionName, List<ColumnStatisticsObj> statistics)
+    public void setPartitionColumnStatistics(String databaseName, String tableName, String partitionName, List<ColumnStatisticsObj> statistics, AcidTransaction transaction)
             throws TException
     {
         ColumnStatisticsDesc statisticsDescription = new ColumnStatisticsDesc(false, databaseName, tableName);
+        ColumnStatistics columnStatistics = new ColumnStatistics(statisticsDescription, statistics);
         statisticsDescription.setPartName(partitionName);
-        ColumnStatistics request = new ColumnStatistics(statisticsDescription, statistics);
-        client.update_partition_column_statistics(request);
+        if (transaction.isAnalyze()) {
+            SetPartitionsStatsRequest request = new SetPartitionsStatsRequest();
+            request.setColStats(ImmutableList.of(columnStatistics));
+            request.setNeedMerge(false);
+            request.setWriteId(transaction.getWriteId());
+            request.setValidWriteIdList(getTableWriteIds(databaseName, tableName, transaction.getAcidTransactionId()));
```

Comment:
I guess this information is used on the metastore side to determine races between ANALYZE and INSERT.

Shouldn't we collect write Ids at the beginning of ANALYZE?

Reply:
Indeed, it's pretty awful to fetch the validWriteIdList for every partition.

I changed the code to fetch the validWriteIdList at the same time the writeId is fetched, and stashed the validWriteIdList in the AcidTransaction object.

Reply:
Did you push this change out?

Reply:
No, after spending many hours trying to get it to work, I reverted the change to pre-cache validWriteIds, as explained in https://github.com/trinodb/trino/pull/7010#issuecomment-792810890.  Also Hive fetches validWriteIds for each partition updated, so not caching matches Hive's behavior.

Reply:
We still need a code-level comment, not to return to this discussion later, and to save someone many hours trying to fix an 'apparent bug'

Reply:
Excellent point.  I added this comment, in setTableStatistics() and setPartitionStatistics():

```
            // Valid writeIds are fetched here rather than at the start of the transaction.
            // I believe the reason we can't fetch at the start of the transaction is that
            // Hive metastore insists that the table or partition be locked for exclusive access,
            // and that only happens when the transaction is being committed.  Moreover, fetching
            // at this point matches Hive behavior - - Hive always fetches valid writeIds for each
            // table or partition updated leading up to committing the transaction.
```